manager/api/transport_test.go

```diff
@@ -0,0 +1,535 @@
+	invalidUser     = manager.User{"userexample.com", "password"}
+	nonexistentUser = manager.User{"non-existentuser@example.com", "pass"}
+	validClient     = manager.Client{Type: "app", Name: "test_app", Payload: "test_payload"}
+	invalidClient   = manager.Client{Type: "foo", Name: "invalid_client", Payload: "some_payload"}
+	validChannel    = manager.Channel{Name: "test"}
+)
+
+func newService() manager.Service {
+	users := mocks.NewUserRepository()
+	clients := mocks.NewClientRepository()
+	channels := mocks.NewChannelRepository()
+	hasher := mocks.NewHasher()
+	idp := mocks.NewIdentityProvider()
+
+	return manager.New(users, clients, channels, hasher, idp)
+}
+
+func runService(test func(manager.Service, string, *http.Client)) {
+	svc := newService()
+	mux := api.MakeHandler(svc)
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+	client := ts.Client()
+
+	test(svc, ts.URL, client)
+}
+
+func sendRequest(client *http.Client, method string, url string, contentType string, token string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+	if token != "" {
+		req.Header.Set("Authorization", token)
+	}
+	if contentType != "" {
+		req.Header.Set("Content-Type", contentType)
+	}
+	return client.Do(req)
+}
+
+func TestRegister(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		validBody, _ := json.Marshal(validUser)
+		invalidBody, _ := json.Marshal(invalidUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+		}{
+			{"register new user", string(validBody), http.StatusCreated},
+			{"try to register existing user", string(validBody), http.StatusConflict},
+			{"try to register user with invalid email address", string(invalidBody), http.StatusBadRequest},
+			{"try to register with invalid request format", "{", http.StatusBadRequest},
+		}
+
+		for _, tc := range cases {
+			res, err := sendRequest(client, http.MethodPost, fmt.Sprintf("%s/users", url), contentType, "", strings.NewReader(tc.req))
+			assert.Nil(t, err, fmt.Sprintf("%s: unexpected error %s", tc.desc, err))
+			assert.Equal(t, tc.status, res.StatusCode, fmt.Sprintf("%s: expected status code %d got %d", tc.desc, tc.status, res.StatusCode))
+		}
+	})
+}
+
+func TestLogin(t *testing.T) {
+	runService(func(svc manager.Service, url string, client *http.Client) {
+		token, _ := json.Marshal(map[string]string{"token": validUser.Email})
+		validCredentials, _ := json.Marshal(validUser)
+		invalidEmailCredentials, _ := json.Marshal(invalidUser)
+		invalidCredentials, _ := json.Marshal(invalidPassUser)
+		nonexistentCredentials, _ := json.Marshal(nonexistentUser)
+		svc.Register(validUser)
+
+		cases := []struct {
+			desc   string
+			req    string
+			status int
+			res    string
+		}{
+			{"login with valid credentials", string(validCredentials), http.StatusCreated, string(token)},
```

Comment:
These `string(...)` are hard to read. Don't "torture" the table, do the cast during test data preparation. This applies to each of the further occurrences.